File: core/src/main/java/heast/core/network/packets/c2s/login/KeyC2SPacket.java
package heast.core.network.packets.c2s.login;

import heast.core.network.Packet;
import heast.core.network.PacketBuf;
import heast.core.network.listeners.ServerLoginListener;
import heast.core.security.RSA;

import javax.crypto.SecretKey;
import java.security.PublicKey;

public final class KeyC2SPacket implements Packet<ServerLoginListener> {

    /**
     * The encrypted secret key.
     */
    final byte[] key;

    public KeyC2SPacket(SecretKey secretKey, PublicKey publicKey) {
        // Manually encrypt the secret AES key, until automatic encryption is activated
        this.key = RSA.encrypt(publicKey, secretKey.getEncoded());
    }

    public KeyC2SPacket(PacketBuf buf) {
        this.key = buf.readByteArray();
    }

    @Override
    public void write(PacketBuf buf) {
        buf.writeByteArray(key);
    }

    @Override
    public void apply(ServerLoginListener listener) {
        listener.onKey(this);
    }

    public byte[] getKey() {
        return key;
    }
}

File: Core/Network/Packets/C2S/KeyC2SPacket.cs
using Core.Network.Listeners;

namespace Core.Network.Packets.C2S;

public class KeyC2SPacket : IPacket
{

	public KeyC2SPacket(byte[] key, byte[] iv)
	{
		Key = key;
		Iv = iv;
	}

	public KeyC2SPacket(PacketBuf buffer)
	{
		Key = buffer.ReadByteArray();
		Iv = buffer.ReadByteArray();
	}

	public byte[] Key { get; }
	public byte[] Iv { get; }

	public void Write(PacketBuf buf)
	{
		buf.WriteByteArray(Key);
		buf.WriteByteArray(Iv);
	}

	public void Apply(IPacketListener listener)
	{
		if (listener is IServerHandshakeListener handshakeListener)
		{
			handshakeListener.OnKey(this);
		}
	}
}